feat: CNPG deployed namespaced

maxlengdell opened the pull request:
CNPG currently relies on the access to cluster nodes. Such ClusterRole should be limited and avoided to harden the CNPG operator deployment. CNPG still relies on webhooks, but that will be addressed in a separate feature request.

With the namespaced deployment of CNPG, the early switchover during a node drain will not be possible. Meaning if someone does a node drain where the primary is deployed, the primary will become unavailable and a replica will be promoted. This is slower than the automatically initiated promotion, but can be mitigated with PodDistruptionBudget settings and topology deployment constraints. So it would be a reasonable trade-off, which should be documented. 

Feature request: #9346

## Changes

Only watch Nodes and ClusterImageCatalog when deployed with namespaced disabled (default)

Signed-off-by: Max Lengdell <max.a.lengdell@ericsson.com>

maxlengdell committed:
feat: CNPG deployed namespaced

Source code additions for making CNPG namespaced.
CNPG currently relies on the access to cluster nodes.
Such ClusterRole should be limited and avoided to harden the CNPG
operator deployment. CNPG still relies on webhooks, but that will be
addressed in a separate feature request.

Feature request: cloudnative-pg#9346

Signed-off-by: Max Lengdell <max.a.lengdell@ericsson.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

maxlengdell committed:
feat: test and doc CNPG deployed namespaced

Test code additions for making CNPG namespaced
This adds a simple upgrade test to enable namespaced deployment of the
operator. It will modify the rbac settings to not allow "nodes" and
"clusterImageCatalog" access. This commit adds documentation details for
the namespaced deployment.

It also adds a new e2e test suite for namespaced_deployment_test which
verifies its possible to deploy. It also verifies the existing
dependency on nodes during node drain is removed.
Namespaced deployment while using PDB during a node drain could reject
the drain. This impact is documented in the chapter about namespaced
deployment

Feature request cloudnative-pg#9346

Signed-off-by: Max Lengdell <max.a.lengdell@ericsson.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

maxlengdell committed:
feat: code review fixes

This commit contains code review corrections. Moving the namespaced
logic from utils to configuration package.
Added a validation function for the configuration. This validation
function can be extended to validate that the configuration provided via
env vars, secrets and configmaps are compatible with each other.

Signed-off-by: Max Lengdell <max.a.lengdell@ericsson.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

maxlengdell committed:
feat: code review corrections

Additional code review corrections to standardize the configuration
error messages. Improve variable names and readability

Signed-off-by: Max Lengdell <max.a.lengdell@ericsson.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

maxlengdell committed:
feat: Add description to error types

Add error type description

Signed-off-by: Max Lengdell <max.a.lengdell@ericsson.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

maxlengdell committed:
feat: Fix test cleanup

Restore operator namespace after the namespaced tests are done.
Remove the upgrade test that was added in previous commit since it does
not really add any extra coverage compared to namespaced_deployment.

Signed-off-by: Max Lengdell <max.a.lengdell@ericsson.com>
Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
chore: obey linter

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
fix: review

- Better errors
- Handle topology (fix)

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
feat(webhook): forbid syncReplicaElectionConstraint in namespaced mode

Added validation in cluster webhook to prevent enabling syncReplicaElectionConstraint when operator is in namespaced mode

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
feat(webhook): forbid ClusterImageCatalog in namespaced mode

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
chore: simplify

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

maxlengdell committed:
fix: fix namespaced test

Fix namespaced tests so they wait properly for the deployment
The linter suggests gingo imports to not be done with a ., but the
existing design pattern seems to use that. Using the existing pattern
for now

Signed-off-by: Max Lengdell <max.a.lengdell@ericsson.com>

maxlengdell commented:
I see the label to backport is there. I do not think this feature qualifies to be backported. Cant remove the label, so hopefully this answers the question in case it pops up

jsilvela commented on docs/src/operator_conf.md:
```diff
@@ -201,4 +202,66 @@ You can also access pprof using the browser at [http://localhost:6060/debug/ppro
+### Known Limitations with PodDisruptionBudget
+
+!!! Warning
+    When running in namespaced mode with PodDisruptionBudget (PDB) enabled, node drain operations may fail.
```
New admonition syntax
```suggestion
:::warning
    When running in namespaced mode with PodDisruptionBudget (PDB) enabled, node drain operations may fail.
:::
```

jsilvela commented on docs/src/operator_conf.md:
````diff
@@ -201,4 +202,66 @@ You can also access pprof using the browser at [http://localhost:6060/debug/ppro
+  enablePDB: false  # Disable PDB in namespaced mode
+  storage:
+    size: 1Gi
+```
````
Not sure having this YAML is worth it. You could easily explain to set `.spec.enablePDB` to false.

jsilvela commented on tests/utils/namespaced/doc.go:
```diff
@@ -0,0 +1,2 @@
+// Package namespaced provides utilities for testing CloudNativePG operator in namespaced deployment mode
```
Missing the header


```suggestion
/*
Copyright © contributors to CloudNativePG, established as
CloudNativePG a Series of LF Projects, LLC.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
*/

// Package namespaced provides utilities for testing CloudNativePG operator in namespaced deployment mode
```

jsilvela reviewed (changes requested):
It's good overall, but I've left a bunch of comments, mostly on documentation and the use of  Context within the Ginkgo tests.

maxlengdell committed:
fix: code review updates

    Fixing minor test changes and documentation updates
    One minor change left is to address the comments in
    cluster_status_test.go

Signed-off-by: Max Lengdell <max.a.lengdell@ericsson.com>

maxlengdell committed:
Merge branch 'main' into dev/9346

Signed-off-by: maxLengdell <max.a.lengdell@ericsson.com>

maxlengdell committed:
fix: code review updates

Fix comments so cluster_status_test getPodsTopology is in its own
describe bracket

Signed-off-by: Max Lengdell <max.a.lengdell@ericsson.com>

jsilvela reviewed (approved):
One optional comment on the use of context in the E2E.
Another one, a suggestion but not a blocker, is to add some content in the document `docs/src/e2e.md`. Flaunt your E2E!
EDIT: come to think of it, you should also strongly consider adding some text in the `security.md` file.

maxlengdell committed:
fix: address optional code reviews

Fix the context comment in the namespaced test.
Add some description to the cluster-role dependencies when namespaced
mode is enabled.

Signed-off-by: Max Lengdell <max.a.lengdell@ericsson.com>

maxlengdell commented:
Hi @sxd, I see your point on that "namespaced deployment" already exist. The PR was probably misdirected. Indeed the only goal for this PR is to make the operator work without dependency on rbac to nodes or ClusterImageCatalog. Agree with your points in the rest of the comment as well.

I'll create 3 new pull requests: 
1. Make CNPG operator run and work without crashing when the SA does not have access to the nodes. To start; I'll make it configurable by a parameter, after that maybe we can find a auto-detect logic. To be discussed in that separate PR. 
2. Make CNPG operator run and work without crashing when the SA does not have access to the ClusterImageCatalog. Same assumption here, making it configurable. I'll do my best to come up with suitable names (the hardest part)

Glad you liked the e2e tests. I'll port those to the separate PRs. This PR can stay open to preserve the state for now and once the new PRs are open we can take it from there if that's ok 😃 

I'll update the feature request so that "namespaced deployment" isnt used, since indeed that's already existing. Thanks for the review!

mnencia commented:
Closing this in favor of #10135, which carries forward the Node-watching work (and the runtime RBAC-splitting package) split out of here, as we discussed above. The "namespaced" framing is being dropped since namespaced operation already exists through `OPERATOR_NAMESPACE`/`WATCH_NAMESPACE`; the remaining piece, running without `ClusterImageCatalog` access, can come as its own opt-in PR.

I know you'd asked to keep this open as a reference until the split PRs landed. It's been a few months and the branch has since drifted into merge conflicts, so I think it's cleaner to close it and track the work in #10135. The diff and the e2e tests stay fully visible here for the ClusterImageCatalog follow-up. Thanks for the work, the e2e coverage especially.
